### PR TITLE
feat: refactor for testability, provide an autofix for material imports rule

### PR DESCRIPTION
**Pull request by jolexxa**
- Refactor for testability
- Provide an autofix for material imports rule
- Rename `url` to `documentationUrl` for clarity.

Regarding the last change with `documentationUrl`, you won't want to merge this in until you pin any repos using this to the specific git commit prior to this one, as this is a breaking change.

If this is too much, we can just manually grab the new functionality (auto fix for material imports) and throw that on another PR. I thought it might be nice to have some tests and clean things up a bit before we added some new rules.

To see test coverage, you can run this snippet. Dart-only projects can be a bit painful to get test coverage in a human-readable way.

```sh
dart test --coverage=./coverage --test-randomize-ordering-seed=random && dart pub global run coverage:format_coverage --lcov --in=coverage --out=coverage/lcov.info --report-on=lib && genhtml coverage/lcov.info -o coverage/ && open coverage/index.html
```

### CODE REVIEW

File: lib/src/analyzer/rules/no_material_cupertino_imports/no_material_cupertino_imports_visitor.dart [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  static const String widgets = 'package:flutter/widgets.dart';
+
+  @override
+  void visitImportDirective(ImportDirective node) {
+    final literal = node.uri;
+    final uri = literal.toSource();
+    final importsMaterial = uri.contains(material);
+    final importsCupertino = uri.contains(cupertino);
+    var fix = uri.replaceAll(material, widgets);
+    fix = fix.replaceAll(material, widgets);
```
Comment (aikalant) [inline lib/src/analyzer/rules/no_material_cupertino_imports/no_material_cupertino_imports_visitor.dart]: Something seems off about these statements - does it work for both `material` and `cupertino`?

File: lib/src/analyzer/options/options.dart [hunk 2]
```diff
@@ -94,10 +99,25 @@ void _applyOptionsToConfig(
...
+  /// Creates a new rule configuration with the given properties.
+  RuleConfig({this.documentationUrl});
+
+  /// Whether or not the rules in this package are enabled.
   bool enabled = true;
-  String? url;
+
+  /// {@template documentation_url}
+  /// Optional path to a markdown file describing the rule. The markdown file
+  /// must contain a heading with the same name as the rule id.
```
Comment (aikalant) [inline lib/src/analyzer/options/options.dart]:
```suggestion
  /// should contain a heading with the same name as the rule id.
```
Technically if the heading does not exist, the web page will just stay scrolled at the top, so not strictly required.

File: lib/src/analyzer/rules/no_material_cupertino_imports/no_material_cupertino_imports_visitor.dart (hunk 1 above)
Comment (jolexxa) [reply lib/src/analyzer/rules/no_material_cupertino_imports/no_material_cupertino_imports_visitor.dart]: Good catch, line 26 should be `fix = fix.replaceAll(cupertino, widgets);`

**Comment by aikalant** [comment]
Edit fixes for the imports rule works great! Having an issue with bare strings getting flagged now though, checked with the base repo and it is working there. Seems to be a problem with the rule itself, not the plugin bit, so I would suggest running the analyzer script via CLI and using the debugger.

**Comment by jolexxa** [comment]
Turns out the mixin you had was a great approach 😉 Didn't realize it, but I had accidentally changed from a recursive visitor to a simple visitor (which doesn't visit all nodes), and that's why the rule had stopped working. Updated the rule visitor hierarchy to more closely resemble what you had before, but with distinct classes to extend for ease of use. Should be good now! 🤞

**Review by aikalant — APPROVED** [review]

**Pull request merged**